### DISCUSSION TITLE
Nested preprocessors bug?

### DISCUSSION THREAD

**Discussion by roozbehid-ic** [dotnet/csharplang#1867] [Language Ideas] [answered]
It seems this piece of code is invalid in C# (which should not be!)

    #if WIN32
          asasdasd ; asd asd  ;asdasd 
           #pragma once
           #define SOMETHING 1
    #endif

The compiler generates 

       CS1025: Single-Line comment or end-of-line expected

note that error is generated for 

     #define SOMETHNG 1

I need this behaviour in my application as I generate cs# files that are valid in both c++ and c#
so I need that #define to be valid for me in the case of c++.

**Comment by svick** [comment, marked as answer]
I think this behaves according to [the specification](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#conditional-compilation-directives) (emphasis mine):

> The remaining *conditional_section*s, if any, are processed as *skipped_section*s: except for pre-processing directives, the source code in the section need not adhere to the lexical grammar; no tokens are generated from the source code in the section; and **pre-processing directives in the section must be lexically correct** but are not otherwise processed.

Since `#define SOMETHING 1` is not a lexically correct pre-processing directive in C#, the code is correctly reported as invalid C# by the compiler.

Since I don't think being able to write source that is valid for both C++ and C# is a goal of C#, you would probably need to find a different reason to change this.

Can't you use different syntax to achieve the same effect, like `const int SOMETHNG = 1;`?

**Comment by SilentSin** [comment]
Times like this it would be nice to have some sort of #IfThisCompilesThenGreatButOtherwiseJustIgnoreIt, but I can't imagine a proposal like that getting much traction.

**Comment by jho1965us** [comment]
Did you consider generating C++/cli instead of C#?
